consensus/istanbul/backend/announce.go

```diff
@@ -544,11 +544,18 @@ func (sb *Backend) generateEncryptedEnodeURLs(enodeURL string, queryEnodeDestAdd
 }
 
 // This function will handle a queryEnode message.
-func (sb *Backend) handleQueryEnodeMsg(peer consensus.Peer, payload []byte) error {
+func (sb *Backend) handleQueryEnodeMsg(addr common.Address, peer consensus.Peer, payload []byte) error {
 	logger := sb.logger.New("func", "handleQueryEnodeMsg")
 
 	msg := new(istanbul.Message)
 
+	// Since this is a gossiped messaged, mark that the peer gossiped it (and presumably processed it) and check to see if this node already processed it
+	sb.markMessageProcessedByPeer(addr, payload)
+	if sb.checkIfMessageProcessedBySelf(payload) {
+		return nil
+	}
+	defer sb.markMessageProcessedBySelf(payload)
```

Comment:
We might as well just call this here instead of deferring to avoid the chance of this message being gossipped to this node from a different peer, right? Same goes for `handleVersionCertificatesMsg`.

One interesting thing we should think about is if we are calling `markMessageProcessedBySelf` (either deferred or not deferred), is it possible that we could receive a message from someone who just joined the validator conn set but this node has not updated the cached validator conn set and then rejects the message at `if !validatorConnSet[msg.Address] {`?

Also if we want to be extra safe, because the announce handler functions are called in a goroutine we may want to add mutexes to these `markMessage*` functions. As far as I can tell there isn't too bad of a worst case if a race condition happens though

Reply:
> We might as well just call this here instead of deferring to avoid the chance of this message being gossipped to this node from a different peer, right? Same goes for `handleVersionCertificatesMsg`

I don't feel too strongly about doing a defer vs executing it here.  The reason I added the defer is so that we it's marked "processed" when it actually is processed.


> One interesting thing we should think about is if we are calling `markMessageProcessedBySelf` (either deferred or not deferred), is it possible that we could receive a message from someone who just joined the validator conn set but this node has not updated the cached validator conn set and then rejects the message at `if !validatorConnSet[msg.Address] {`?

Ya, great point.   In general, I see that there is a high level tradeoff between how aggressively we want these gossip messages to be propagated throughout the network and how much safeguards we want to put in to prevent an attacker to flood the network.  E.g. in the `if !validatorConnSet[msg.Address] {` case, that obviously is needed so that a non validator can't flood the network, but to your point, there is a small chance that enough validators are not synced up and the gossip message is not communicated to the whole network.

I think that the only solution to that is to make sure that retry messages are sent once in a while (similar to the retry of round change messages in the consensus code).   And this is done by the announce thread in that it will periodically send out a version message.